Sources/NIO/LengthFieldBasedFrameDecoder.swift

```diff
@@ -0,0 +1,61 @@
+//===----------------------------------------------------------------------===//
+//
+// This source file is part of the SwiftNIO open source project
+//
+// Copyright (c) 2017-2018 Apple Inc. and the SwiftNIO project authors
+// Licensed under Apache License v2.0
+//
+// See LICENSE.txt for license information
+// See CONTRIBUTORS.txt for the list of SwiftNIO project authors
+//
+// SPDX-License-Identifier: Apache-2.0
+//
+//===----------------------------------------------------------------------===//
+
+/// A `ChannelInboundHandler` used to decode length-delimited frames.
+///
+/// This is a protocol that has messages prefixed with an encoding of their length.
+/// Such a protocol would look like this:
+/// ````
+///     <fixed width integer n in network byte order><n bytes>
+/// ````
+/// ````
+///  BEFORE DECODE (14 bytes)         AFTER DECODE (12 bytes)
+/// +--------+----------------+      +----------------+
+/// | Length | Actual Content |----->| Actual Content |
+/// | 0x000C | "HELLO, WORLD" |      | "HELLO, WORLD" |
+/// +--------+----------------+      +----------------+
+/// ````
+public final class LengthFieldBasedFrameDecoder<T: FixedWidthInteger>: ByteToMessageDecoder {
+    public typealias InboundIn = ByteBuffer
+    public typealias InboundOut = ByteBuffer
+    
+    public var cumulationBuffer: ByteBuffer?
+    
+    private var state: State = .len
+    
+    enum State {
+        case len
```

Comment:
I'd argue for making those cases more descriptive. Like `case waitingForLength` and `case waitingForPayload(Int)`. @normanmaurer / @Lukasa  what do you think?